# Importë¥¼ ê·¸ë£¹í™”í•˜ì—¬ ì •ë¦¬í•©ë‹ˆë‹¤ (`hardsort/import-grouping`)

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

í•´ë‹¹ ë¬¸ì„œì—ì„œëŠ” `import-grouping` ê·œì¹™ì˜ ê¸°ë³¸ ì„¤ì •ê°’ì„ ê¸°ì¤€ìœ¼ë¡œ ì†Œê°œí•˜ë©°, `sortedOrders`ë¥¼ ë¹„ë¡¯í•˜ì—¬ ë‹¤ë¥¸ ì˜µì…˜ì— ëŒ€í•œ ìì„¸í•œ ë‚´ìš©ì€ [ì—¬ê¸°](https://github.com/Sotaneum/eslint-plugin-hardsort/wiki/options)ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”.

## Rule Details

`import-grouping` ê·œì¹™ì˜ ê¸°ë³¸ ì •ë ¬ì€ ë‹¤ìŒê³¼ ê°™ìŠµë‹ˆë‹¤.

```
1. defaultì¸ ê²ƒê³¼ ì•„ë‹Œ ê²ƒ
2. aliasë¥¼ ì‚¬ìš©í•˜ëŠ” ê²ƒê³¼ ì•„ë‹Œ ê²ƒ
```

### defaultì¸ ê²ƒê³¼ ì•„ë‹Œ ê²ƒ

ë³„ë„ì˜ ì˜µì…˜ì´ ì—†ëŠ” ê²½ìš° ëª¨ë“  importëŠ” defaultê·¸ë£¹ìœ¼ë¡œ ì·¨ê¸‰ë©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
import React from "react";

import Button from "@/components/Button";
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
import React from "react";
import Button from "@/components/Button";
```

### aliasë¥¼ ì‚¬ìš©í•˜ëŠ” ê²ƒê³¼ ì•„ë‹Œ ê²ƒ

ê°™ì€ ê·¸ë£¹ê°„ì— ì •ë ¬ë„ ì§€ì •í•  ìˆ˜ ìˆìŠµë‹ˆë‹¤. alias ì‚¬ìš©ì‹œ ì•„ë˜ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
importCheckbox from "./Checkbox";
import Button from "@/components/Button";
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
import Button from "@/components/Button";
importCheckbox from "./Checkbox";
```

### Options

#### grouping

ê·¸ë£¹ì„ ì‹ë³„í•  ìˆ˜ ìˆëŠ” `id`, í•´ë‹¹ ê·¸ë£¹ì— ì†í•  `pathPatterns`, ê·¸ë£¹ ë‚´ì˜ ì •ë ¬ ë°©ì‹ì„ ì§€ì •í•˜ëŠ” `sortedOrders` ë¥¼ ì§€ì •í•©ë‹ˆë‹¤.
ë°°ì—´ë¡œ ì§€ì •í•˜ë©°, ë¨¼ì € ì •ì˜í•œ ë‚´ìš©ë¶€í„° ë¶„ë°°ë©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  {
    "id": "default",
    "sortedOrders": ["aliasDesc"]
  }
]
```

**defaultëŠ” pathPatternsì„ ì§€ì •í•˜ì—¬ë„ ë‚˜ë¨¸ì§€ ì •ë³´ê°€ í¬í•¨ë©ë‹ˆë‹¤.**

#### sort

`alias`, `orders`ë¥¼ í†µí•´ alias ì´í•´ì™€ ê° ê·¸ë£¹ì„ í‘œì‹œë  ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
{
  "orders": ["default"]
}
```

## When Not To Use It

í•´ë‹¹ ê·œì¹™ì€ import êµ¬ë¶„ì´ í•­ìƒ íŒŒì¼ì˜ ìƒë‹¨ì—ë§Œ ìœ„ì¹˜í•œë‹¤ëŠ” ê°€ì •ì„ í•˜ê³  ìˆìŠµë‹ˆë‹¤.
ë”°ë¼ì„œ, import ìƒë‹¨ì— ë‹¤ë¥¸ êµ¬ë¬¸ì´ ìˆëŠ” ê²½ìš° ì‚¬ìš©ì„ í”¼í•´ì•¼í•©ë‹ˆë‹¤.
