# JSX Attributesë¥¼ ì •ë ¬í•©ë‹ˆë‹¤ (`hardsort/jsx-attr-sort`)

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

í•´ë‹¹ ë¬¸ì„œì—ì„œëŠ” `jsx-attr-sort` ê·œì¹™ì˜ ê¸°ë³¸ ì„¤ì •ê°’ì„ ê¸°ì¤€ìœ¼ë¡œ ì†Œê°œí•˜ë©°, `sortedOrders`ë¥¼ ë¹„ë¡¯í•˜ì—¬ ë‹¤ë¥¸ ì˜µì…˜ì— ëŒ€í•œ ìì„¸í•œ ë‚´ìš©ì€ [ì—¬ê¸°](https://github.com/Sotaneum/eslint-plugin-hardsort/wiki/options)ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”.

## Rule Details

`jsx-attr-sort` ê·œì¹™ì˜ ê¸°ë³¸ ì •ë ¬ì€ ë‹¤ìŒê³¼ ê°™ìŠµë‹ˆë‹¤.

```
1. boolean íƒ€ì… ìƒëµ ì—¬ë¶€
2. í•¨ìˆ˜ ì—¬ë¶€
3. keyì˜ ê¸¸ì´ ìˆœì„œ
4. keyì˜ ì•ŒíŒŒë²³ ìˆœì„œ
```

### Boolean íƒ€ì… ìƒëµ ì—¬ë¶€

Boolean íƒ€ì…ì¸ attributeì—ì„œ ìƒëµí–ˆì„ ê²½ìš° ì•ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
<Foo onClick={() => {}} title="Download" check />
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
<Foo check title="Download" onClick={() => {}} />
```

### í•¨ìˆ˜ ì—¬ë¶€

í•¨ìˆ˜ê°€ ì•„ë‹Œ ë³€ìˆ˜ë¥¼ ì•ìª½ì— ìœ„ì¹˜í•˜ê³  í•¨ìˆ˜ì¸ ë³€ìˆ˜ë¥¼ ë’¤ìª½ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
<Foo onClick={handleClick} cry={name} />
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
<Foo cry={name} onClick={handleClick} />
```

### keyì˜ ê¸¸ì´ ìˆœì„œ

keyì˜ ê¸¸ì´ì— ë”°ë¼ ì§§ì€ ìˆœì„œê°€ ì•ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
<Foo charm climb cry capture cake />
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
<Foo cry cake charm climb capture />
```

### keyì˜ ì•ŒíŒŒë²³ ìˆœì„œ

keyì˜ ì•ŒíŒŒë²³ ìˆœì„œì— ë”°ë¼ ì•ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
<Foo love bold kind care open />
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
<Foo bold care kind love open />
```

### Spread ê·¸ë£¹

`Spread`ë¡œ `attribute` ê°’ì„ ì „ë‹¬í•˜ê³  ìˆëŠ” ê²½ìš°, `Spread` ê¸°ì¤€ìœ¼ë¡œ ê·¸ë£¹í™”í•˜ì—¬ ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
<Foo
  capture=""
  climb=""
  charm=""
  {...props}
  onClick={handleClick}
  cake=""
  cry={name}
/>
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
<Foo
  charm=""
  climb=""
  capture=""
  {...props}
  cry={name}
  cake=""
  onClick={handleClick}
/>
```

### Options

#### sortedOrders

ì •ë ¬ ë°©ì‹ ë° ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "isShortAsc",
  "isFunctionAsc",
  "keyNameLengthAsc",
  "localeCompareAsc",
]
```

#### functionKeyNames

í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” ì´ë¦„ ëª©ë¡ì„ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "next",
  "select",
]
```

#### ignoreEnterGroup

ì—”í„°ë¥¼ ê¸°ì¤€ìœ¼ë¡œ ê·¸ë£¹ì„ êµ¬ì„±í•˜ì§€ ì•Šì„ ì§€ ì—¬ë¶€ë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
true
```

#### functionValueTypes

valueê°€ ìˆëŠ” ê²½ìš°, í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” valueì˜ typeë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "AsyncFunction",
  "MethodDefinition",
  "GeneratorFunction",
  "FunctionExpression",
  "ArrowFunctionExpression",
]
```

#### functionKeyNamePatterns

í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” ì´ë¦„ íŒ¨í„´ ëª©ë¡ì„ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "^on",
  "^callback",
]
```

## When Not To Use It

- ì´ë¯¸ `JSX Attribute` ì •ë ¬ê³¼ ê´€ë ¨ëœ í”ŒëŸ¬ê·¸ì¸ ë° ê·œì¹™ì„ ì‚¬ìš©í•˜ëŠ” ê²½ìš°, ì¶©ëŒì´ ë°œìƒí•  ìˆ˜ ìˆìœ¼ë¯€ë¡œ, í•„ìš”í•œ ê²½ìš°ì— ì‚¬ìš©í•˜ì„¸ìš”.
