# Objectë¥¼ ì •ë ¬í•©ë‹ˆë‹¤ (`hardsort/object-sort`)

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

í•´ë‹¹ ë¬¸ì„œì—ì„œëŠ” `object-sort` ê·œì¹™ì˜ ê¸°ë³¸ ì„¤ì •ê°’ì„ ê¸°ì¤€ìœ¼ë¡œ ì†Œê°œí•˜ë©°, `sortedOrders`ë¥¼ ë¹„ë¡¯í•˜ì—¬ ë‹¤ë¥¸ ì˜µì…˜ì— ëŒ€í•œ ìì„¸í•œ ë‚´ìš©ì€ [ì—¬ê¸°](https://github.com/Sotaneum/eslint-plugin-hardsort/wiki/options)ë¥¼ ì°¸ì¡°í•˜ì„¸ìš”.

## Rule Details

`object-sort` ê·œì¹™ì˜ ê¸°ë³¸ ì •ë ¬ì€ ë‹¤ìŒê³¼ ê°™ìŠµë‹ˆë‹¤.

```
1. í•¨ìˆ˜ ì—¬ë¶€
2. ì¶•ì•½ì´ ì•„ë‹ˆê±°ë‚˜ ê¸°ë³¸ê°’ì´ ìˆê±°ë‚˜ ìƒˆë¡œìš´ ì´ë¦„ìœ¼ë¡œ í• ë‹¹ ì—¬ë¶€
3. keyì˜ ê¸¸ì´ ìˆœì„œ
4. keyì˜ ì•ŒíŒŒë²³ ìˆœì„œ
```

### í•¨ìˆ˜ ì—¬ë¶€

í•¨ìˆ˜ê°€ ì•„ë‹Œ ë³€ìˆ˜ë¥¼ ì•ìª½ì— ìœ„ì¹˜í•˜ê³  í•¨ìˆ˜ì¸ ë³€ìˆ˜ë¥¼ ë’¤ìª½ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```js
// âœ— bad
const foo = { onClick, title };
```

Examples of **correct** code for this rule:

```js
// âœ“ good
const foo = { title, onClick };
```

### ì¶•ì•½ì´ ì•„ë‹ˆê±°ë‚˜ ê¸°ë³¸ê°’ì´ ìˆê±°ë‚˜ ìƒˆë¡œìš´ ì´ë¦„ìœ¼ë¡œ í• ë‹¹ ì—¬ë¶€

ì¶•ì•½ì´ ì•„ë‹ˆê±°ë‚˜ ê¸°ë³¸ê°’ì´ ìˆê±°ë‚˜ ìƒˆë¡œìš´ ì´ë¦„ìœ¼ë¡œ í• ë‹¹í–ˆì„ ê²½ìš° ë’¤ìª½ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
const foo = { done: () => {}, context, onClick, title: "Download" };
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
const foo = { context, title: "Download", onClick, done: () => {} };
```

### keyì˜ ê¸¸ì´ ìˆœì„œ

keyì˜ ê¸¸ì´ì— ë”°ë¼ ì§§ì€ ìˆœì„œê°€ ì•ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
const foo = { charm, climb, cry, capture, cake };
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
const foo = { cry, cake, charm, climb, capture };
```

### keyì˜ ì•ŒíŒŒë²³ ìˆœì„œ

keyì˜ ì•ŒíŒŒë²³ ìˆœì„œì— ë”°ë¼ ì•ì— ìœ„ì¹˜í•˜ë„ë¡ í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
const foo = { care, open, kind, love, bold };
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
const foo = { bold, care, kind, love, open };
```

### Spread ê·¸ë£¹

`Spread`ë¡œ `attribute` ê°’ì„ ì „ë‹¬í•˜ê³  ìˆëŠ” ê²½ìš°, `Spread` ê¸°ì¤€ìœ¼ë¡œ ê·¸ë£¹í™”í•˜ì—¬ ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
const foo = {
  title: "Not Found",
  ...defaultOptions,
  done: () => {},
  check: false,
};
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
const foo = {
  title: "Not Found",
  ...defaultOptions,
  check: false,
  done: () => {},
};
```

### Enter ê·¸ë£¹

`Enter`ë¥¼ ê¸°ì¤€ìœ¼ë¡œ ê·¸ë£¹í™”í•˜ì—¬ ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.
`Enter` ë°”ë¡œ ìœ„ì— ìˆëŠ” Line ì£¼ì„ì€ ê·¸ë£¹ì˜ ì œëª©ìœ¼ë¡œ ì·¨ê¸‰ë©ë‹ˆë‹¤.

Examples of **incorrect** code for this rule:

```jsx
// âœ— bad
module.exports = {
  CODE: {
    // HTTP CODE
    UNAUTHORIZED: "401",
    NOT_FOUND: "404",
    OK: "200",

    // API ERROR CODE
    INVALID: "-1",
  },
};
```

Examples of **correct** code for this rule:

```jsx
// âœ“ good
module.exports = {
  CODE: {
    // HTTP CODE
    OK: "200",
    NOT_FOUND: "404",
    UNAUTHORIZED: "401",

    // API ERROR CODE
    INVALID: "-1",
  },
};
```

### Options

#### sortedOrders

ì •ë ¬ ë°©ì‹ ë° ìˆœì„œë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "isFunctionAsc",
  "isShortAsc",
  "keyNameLengthAsc",
  "localeCompareAsc",
]
```

#### functionKeyNames

í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” ì´ë¦„ ëª©ë¡ì„ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "next",
  "select",
]
```

#### ignoreEnterGroup

ì—”í„°ë¥¼ ê¸°ì¤€ìœ¼ë¡œ ê·¸ë£¹ì„ êµ¬ì„±í•˜ì§€ ì•Šì„ ì§€ ì—¬ë¶€ë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
false
```

#### functionValueTypes

valueê°€ ìˆëŠ” ê²½ìš°, í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” valueì˜ typeë¥¼ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "AsyncFunction",
  "MethodDefinition",
  "GeneratorFunction",
  "FunctionExpression",
  "ArrowFunctionExpression",
]
```

#### functionKeyNamePatterns

í•¨ìˆ˜ë¡œ ê°€ì •í•˜ëŠ” ì´ë¦„ íŒ¨í„´ ëª©ë¡ì„ ì§€ì •í•©ë‹ˆë‹¤.

```
# ê¸°ë³¸ê°’
[
  "^on",
  "^callback",
]
```

## When Not To Use It

- ì´ë¯¸ `Object` ì •ë ¬ê³¼ ê´€ë ¨ëœ í”ŒëŸ¬ê·¸ì¸ ë° ê·œì¹™ì„ ì‚¬ìš©í•˜ëŠ” ê²½ìš°, ì¶©ëŒì´ ë°œìƒí•  ìˆ˜ ìˆìœ¼ë¯€ë¡œ, í•„ìš”í•œ ê²½ìš°ì— ì‚¬ìš©í•˜ì„¸ìš”.
